File: python/design_patterns.py
# design_patterns.py

# Design Patterns in Python

# 1. Singleton Pattern
class Singleton:
    _instance = None

    def __new__(cls, *args, **kwargs):
        if not cls._instance:
            cls._instance = super(Singleton, cls).__new__(cls)
        return cls._instance


# Example usage of Singleton
singleton1 = Singleton()
singleton2 = Singleton()
print("Singleton pattern:", singleton1 is singleton2)  # True


# 2. Factory Pattern
class Dog:
    def speak(self):
        return "Woof!"


class Cat:
    def speak(self):
        return "Meow!"


class AnimalFactory:
    @staticmethod
    def get_animal(animal_type):
        if animal_type == "dog":
            return Dog()
        elif animal_type == "cat":
            return Cat()


# Example usage of Factory
dog = AnimalFactory.get_animal("dog")
cat = AnimalFactory.get_animal("cat")
print("Factory pattern (Dog):", dog.speak())
print("Factory pattern (Cat):", cat.speak())


# 3. Observer Pattern
class Subject:
    def __init__(self):
        self._observers = []

    def attach(self, observer):
        if observer not in self._observers:
            self._observers.append(observer)

    def detach(self, observer):
        try:
            self._observers.remove(observer)
        except ValueError:
            pass

    def notify(self):
        for observer in self._observers:
            observer.update(self)


class ConcreteObserver:
    def update(self, subject):
        print(f"Observer: {subject} has been updated.")


# Example usage of Observer
subject = Subject()
observer1 = ConcreteObserver()
observer2 = ConcreteObserver()

subject.attach(observer1)
subject.attach(observer2)

subject.notify()


# 4. Strategy Pattern
class Strategy:
    def execute(self, data):
        raise NotImplementedError()


class ConcreteStrategyA(Strategy):
    def execute(self, data):
        print(f"Strategy A: {data}")


class ConcreteStrategyB(Strategy):
    def execute(self, data):
        print(f"Strategy B: {data}")


class Context:
    def __init__(self, strategy: Strategy):
        self._strategy = strategy

    def set_strategy(self, strategy: Strategy):
        self._strategy = strategy

    def execute_strategy(self, data):
        self._strategy.execute(data)


# Example usage of Strategy
context = Context(ConcreteStrategyA())
context.execute_strategy("Data for A")
context.set_strategy(ConcreteStrategyB())
context.execute_strategy("Data for B")


# 5. Decorator Pattern
class Coffee:
    def cost(self):
        return 5


class MilkDecorator:
    def __init__(self, coffee):
        self._coffee = coffee

    def cost(self):
        return self._coffee.cost() + 2


class SugarDecorator:
    def __init__(self, coffee):
        self._coffee = coffee

    def cost(self):
        return self._coffee.cost() + 1


# Example usage of Decorator
coffee = Coffee()
print("Base coffee cost:", coffee.cost())

milk_coffee = MilkDecorator(coffee)
print("Coffee with milk cost:", milk_coffee.cost())

sugar_milk_coffee = SugarDecorator(milk_coffee)
print("Coffee with milk and sugar cost:", sugar_milk_coffee.cost())

File: typescript/design_patterns.ts
// Design Patterns in TypeScript

// 1. Singleton Pattern
class Singleton {
  private static instance: Singleton;

  private constructor() {
    // Private constructor prevents instantiation from other classes
  }

  public static getInstance(): Singleton {
    if (!Singleton.instance) {
      Singleton.instance = new Singleton();
    }
    return Singleton.instance;
  }

  public someBusinessLogic() {
    console.log("Executing some business logic.");
  }
}

const singleton1 = Singleton.getInstance();
const singleton2 = Singleton.getInstance();

console.log(singleton1 === singleton2); // Outputs: true

// 2. Factory Pattern
interface Animal {
  speak(): string;
}

class Dog implements Animal {
  speak(): string {
    return "Woof!";
  }
}

class Cat implements Animal {
  speak(): string {
    return "Meow!";
  }
}

class AnimalFactory {
  static createAnimal(animalType: string): Animal {
    switch (animalType) {
      case "dog":
        return new Dog();
      case "cat":
        return new Cat();
      default:
        throw new Error("Unknown animal type");
    }
  }
}

const dog = AnimalFactory.createAnimal("dog");
const cat = AnimalFactory.createAnimal("cat");
console.log(dog.speak()); // Outputs: Woof!
console.log(cat.speak()); // Outputs: Meow!

// 3. Observer Pattern
interface Observer {
  update(message: string): void;
}

class ConcreteObserver implements Observer {
  private observerName: string;

  constructor(name: string) {
    this.observerName = name;
  }

  update(message: string): void {
    console.log(`${this.observerName} received: ${message}`);
  }
}

class Subject {
  private observers: Observer[] = [];

  addObserver(observer: Observer): void {
    this.observers.push(observer);
  }

  removeObserver(observer: Observer): void {
    this.observers = this.observers.filter((obs) => obs !== observer);
  }

  notifyObservers(message: string): void {
    for (const observer of this.observers) {
      observer.update(message);
    }
  }
}

const subject = new Subject();
const observer1 = new ConcreteObserver("Observer 1");
const observer2 = new ConcreteObserver("Observer 2");

subject.addObserver(observer1);
subject.addObserver(observer2);

subject.notifyObservers("Event 1 occurred!"); // Both observers receive the message

// 4. Strategy Pattern
interface PaymentStrategy {
  pay(amount: number): void;
}

class CreditCardPayment implements PaymentStrategy {
  pay(amount: number): void {
    console.log(`Paid ${amount} using Credit Card.`);
  }
}

class PayPalPayment implements PaymentStrategy {
  pay(amount: number): void {
    console.log(`Paid ${amount} using PayPal.`);
  }
}

class PaymentContext {
  private strategy: PaymentStrategy;

  constructor(strategy: PaymentStrategy) {
    this.strategy = strategy;
  }

  setStrategy(strategy: PaymentStrategy): void {
    this.strategy = strategy;
  }

  executeStrategy(amount: number): void {
    this.strategy.pay(amount);
  }
}

const paymentContext = new PaymentContext(new CreditCardPayment());
paymentContext.executeStrategy(100); // Outputs: Paid 100 using Credit Card.

paymentContext.setStrategy(new PayPalPayment());
paymentContext.executeStrategy(200); // Outputs: Paid 200 using PayPal.

// 5. Decorator Pattern
interface Coffee {
  cost(): number;
}

class SimpleCoffee implements Coffee {
  cost(): number {
    return 5;
  }
}

class MilkDecorator implements Coffee {
  private coffee: Coffee;

  constructor(coffee: Coffee) {
    this.coffee = coffee;
  }

  cost(): number {
    return this.coffee.cost() + 2;
  }
}

class SugarDecorator implements Coffee {
  private coffee: Coffee;

  constructor(coffee: Coffee) {
    this.coffee = coffee;
  }

  cost(): number {
    return this.coffee.cost() + 1;
  }
}

let coffee: Coffee = new SimpleCoffee();
console.log(`Cost: $${coffee.cost()}`); // Outputs: Cost: $5

coffee = new MilkDecorator(coffee);
coffee = new SugarDecorator(coffee);
console.log(`Cost with milk and sugar: $${coffee.cost()}`); // Outputs: Cost with milk and sugar: $8

// 6. Command Pattern
class Light {
  turnOn(): void {
    console.log("Light is On");
  }

  turnOff(): void {
    console.log("Light is Off");
  }
}

interface Command {
  execute(): void;
}

class LightOnCommand implements Command {
  private light: Light;

  constructor(light: Light) {
    this.light = light;
  }

  execute(): void {
    this.light.turnOn();
  }
}

class LightOffCommand implements Command {
  private light: Light;

  constructor(light: Light) {
    this.light = light;
  }

  execute(): void {
    this.light.turnOff();
  }
}

class RemoteControl {
  private command: Command;

  setCommand(command: Command): void {
    this.command = command;
  }

  pressButton(): void {
    this.command.execute();
  }
}

const light = new Light();
const lightOn = new LightOnCommand(light);
const lightOff = new LightOffCommand(light);

const remote = new RemoteControl();
remote.setCommand(lightOn);
remote.pressButton(); // Outputs: Light is On

remote.setCommand(lightOff);
remote.pressButton(); // Outputs: Light is Off

// 7. Proxy Pattern
interface SubjectInterface {
  request(): void;
}

class RealSubject implements SubjectInterface {
  request(): void {
    console.log("RealSubject: Handling request.");
  }
}

class ProxySubject implements SubjectInterface {
  private realSubject: RealSubject;

  constructor(realSubject: RealSubject) {
    this.realSubject = realSubject;
  }

  request(): void {
    if (this.checkAccess()) {
      this.realSubject.request();
    }
  }

  private checkAccess(): boolean {
    console.log("Proxy: Checking access before firing a request.");
    return true;
  }
}

const realSubject = new RealSubject();
const proxy = new ProxySubject(realSubject);
proxy.request(); // Outputs: Proxy: Checking access before firing a request. RealSubject: Handling request.
